patches/sitecustomize.py

```diff
@@ -1,22 +1,56 @@
 # TODO(rosbo): Remove this once we fix the issue with fastai importing older libcudnn if imported prior to tensorflow
 import tensorflow
 import os
+from google.auth import credentials
+from google.cloud import bigquery
+from google.cloud.bigquery._http import Connection
 
-# Monkey patches BigQuery client creation to use proxy.
-kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
-if kaggle_proxy_token:
-    from google.auth import credentials
-    from google.cloud import bigquery
-    from google.cloud.bigquery._http import Connection
+class KaggleKernelCredentials(credentials.Credentials):
+    def refresh(self, request):
+        print("Calling Kaggle.UserSecrets to refresh token.")
+        # Set self.token and self.expiry here.
+        raise NotImplementedError("Private BigQuery integration is not yet implemented.")
 
-    Connection.API_BASE_URL = os.getenv("KAGGLE_DATA_PROXY_URL")
-    Connection._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = kaggle_proxy_token
+kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
+CONNECTION_BASE_URL = Connection.API_BASE_URL
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs.get('project').lower() if 'project' in kwargs else None
+    # Use Data Proxy if user has specified to use the Kaggle project, or if
+    # there are no connected GCP accounts (to maintain backwards compatibility).
+    if bq_user_jwt is None and specified_project and specified_project != 'kaggle':
+        raise Exception("In order to query a private BigQuery project, please connect a GCP account. "
+                        "Otherwise specify 'kaggle' as the project to use Kaggle's public dataset BigQuery integration.")
+    use_data_proxy = specified_project == 'kaggle' or bq_user_jwt is None
```

Comment:
Should `'kaggle'` be `'kaggle-161607'`?
  https://github.com/Kaggle/Kaggle.SDS.Worker/blob/720f4aa8ad15549e2c1069c557ccca2480dd9f32/kernels/worker/src/kernel_run_container_commands.py#L62

I'm not sure why you need that test in the first place.  It doesn't exist on the LHS for this case, and it seems preferable not to specify the kaggle project again, nor to expose it more than it already is.

Reply:
This kaggle "project" the user specifies is really more of just a way to tell us they want to use the Kaggle "free bigquery" Data Proxy even though they've connected a GCP account. If they haven't connected their GCP account (ie. bq_user_jwt is not present) then they don't need specify a 'kaggle' or any other project, we will default to using the data proxy. 

More info here: https://github.com/Kaggle/docker-python/pull/414#discussion_r243672377